File: apps/ios/ChalkMeetingKit/Sources/ChalkMeetingKit/ChalkFileLogger.swift
import Foundation

public enum ChalkLogLevel: String {
	case debug
	case info
	case error
}

public final class ChalkFileLogger: @unchecked Sendable {
	public static let shared = ChalkFileLogger()

	private let queue = DispatchQueue(label: "ai.q9labs.chalk.file-logger")
	private var dirUrl: URL?

	private init() {}

	public func configure(directory: URL? = nil) {
		queue.sync {
			if let directory {
				dirUrl = directory
				return
			}

			let base = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first
			dirUrl = base?.appendingPathComponent("chalk-logs", isDirectory: true)
		}
	}

	public func log(_ level: ChalkLogLevel, _ message: String, meta: [String: String] = [:]) {
		queue.async { [weak self] in
			guard let self else { return }
			let dir = self.ensureDir()
			guard let dir else { return }

			let ts = ISO8601DateFormatter().string(from: Date())
			let metaStr = meta.isEmpty
				? ""
				: " " + meta
					.sorted(by: { $0.key < $1.key })
					.map { "\($0.key)=\($0.value)" }
					.joined(separator: " ")
			let line = "\(ts) level=\(level.rawValue)\(metaStr) msg=\(self.escape(message))\n"

			self.append(line, to: dir.appendingPathComponent("chalk.log"))
			if level == .debug { self.append(line, to: dir.appendingPathComponent("chalk.debug.log")) }
			if level == .error { self.append(line, to: dir.appendingPathComponent("chalk.error.log")) }
		}
	}

	public func files() -> [URL] {
		queue.sync {
			let dir = ensureDir()
			guard let dir else { return [] }
			return [
				dir.appendingPathComponent("chalk.log"),
				dir.appendingPathComponent("chalk.debug.log"),
				dir.appendingPathComponent("chalk.error.log"),
			].filter { FileManager.default.fileExists(atPath: $0.path) }
		}
	}

	public func clear() {
		queue.async { [weak self] in
			guard let self else { return }
			let dir = self.ensureDir()
			guard let dir else { return }
			for name in ["chalk.log", "chalk.debug.log", "chalk.error.log"] {
				try? FileManager.default.removeItem(at: dir.appendingPathComponent(name))
			}
		}
	}

	private func ensureDir() -> URL? {
		if dirUrl == nil { configure() }
		guard let dirUrl else { return nil }

		do {
			try FileManager.default.createDirectory(at: dirUrl, withIntermediateDirectories: true)
			return dirUrl
		} catch {
			return nil
		}
	}

	private func append(_ line: String, to url: URL) {
		let maxBytes = 5 * 1024 * 1024
		if let size = (try? FileManager.default.attributesOfItem(atPath: url.path)[.size] as? NSNumber)?.intValue, size > maxBytes {
			try? FileManager.default.removeItem(at: url)
		}

		if !FileManager.default.fileExists(atPath: url.path) {
			try? line.data(using: .utf8)?.write(to: url, options: .atomic)
			return
		}

		guard let fh = try? FileHandle(forWritingTo: url) else { return }
		defer { try? fh.close() }
		do {
			try fh.seekToEnd()
			if let data = line.data(using: .utf8) {
				try fh.write(contentsOf: data)
			}
		} catch {
			// ignore
		}
	}

	private func escape(_ s: String) -> String {
		s
			.replacingOccurrences(of: "\\", with: "\\\\")
			.replacingOccurrences(of: "\n", with: "\\n")
			.replacingOccurrences(of: "\r", with: "\\r")
			.replacingOccurrences(of: "\t", with: "\\t")
	}
}

File: apps/android/meetingkit/src/main/java/ai/q9labs/chalk/meetingkit/ChalkFileLogger.kt
package ai.q9labs.chalk.meetingkit

import android.content.Context
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

enum class ChalkLogLevel {
	DEBUG,
	INFO,
	ERROR,
}

object ChalkFileLogger {
	private var dir: File? = null
	private val lock = Any()
	private val tsFmt = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX", Locale.US)

	fun init(context: Context) {
		synchronized(lock) {
			if (dir != null) return
			val d = File(context.filesDir, "chalk-logs")
			d.mkdirs()
			dir = d
		}
	}

	fun log(level: ChalkLogLevel, msg: String, meta: Map<String, String> = emptyMap(), err: Throwable? = null) {
		val d = synchronized(lock) { dir } ?: return
		val ts = tsFmt.format(Date())
		val metaStr = if (meta.isEmpty()) "" else " " + meta.entries
			.sortedBy { it.key }
			.joinToString(" ") { "${it.key}=${escape(it.value)}" }
		val errStr = if (err == null) "" else " err=${escape(err.stackTraceToString())}"
		val line = "$ts level=${level.name.lowercase()}$metaStr msg=${escape(msg)}$errStr\n"

		append(File(d, "chalk.log"), line)
		if (level == ChalkLogLevel.DEBUG) append(File(d, "chalk.debug.log"), line)
		if (level == ChalkLogLevel.ERROR) append(File(d, "chalk.error.log"), line)
	}

	fun files(): List<File> {
		val d = synchronized(lock) { dir } ?: return emptyList()
		return listOf("chalk.log", "chalk.debug.log", "chalk.error.log")
			.map { File(d, it) }
			.filter { it.exists() }
	}

	fun clear() {
		val d = synchronized(lock) { dir } ?: return
		for (name in listOf("chalk.log", "chalk.debug.log", "chalk.error.log")) {
			runCatching { File(d, name).delete() }
		}
	}

	private fun append(file: File, line: String) {
		val maxBytes = 5 * 1024 * 1024
		runCatching {
			if (file.exists() && file.length() > maxBytes) file.delete()
			file.appendText(line)
		}
	}

	private fun escape(s: String) =
		s.replace("\\", "\\\\")
			.replace("\n", "\\n")
			.replace("\r", "\\r")
			.replace("\t", "\\t")
}

